File: src/main/java/org/fx/debugger/view/AboutDialog.java
package org.fx.debugger.view;

import io.avaje.inject.Lazy;
import it.unimi.dsi.fastutil.objects.Object2ObjectArrayMap;
import it.unimi.dsi.fastutil.objects.Object2ObjectMap;
import jakarta.inject.Singleton;
import java.util.Map;
import java.util.function.Supplier;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.control.ButtonType;
import javafx.scene.control.Dialog;
import javafx.scene.control.Label;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.fx.debugger.constant.ManifestAttribute;
import org.fx.debugger.model.SelfManifest;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

@Singleton
@Lazy
@Slf4j
public class AboutDialog extends Dialog<Void> {

  private final GridPane content = new GridPane();

  {
    content.setPadding(new Insets(20));
    content.setHgap(10);
    content.setVgap(5);
  }

  private final SelfManifest selfManifest;

  {
    getDialogPane().getButtonTypes().add(ButtonType.CLOSE);
    setResizable(true);
    setTitle("About Fx-Debugger");
  }

  private final HBox h = new HBox(content);

  private final VBox v = new VBox(h);

  {
    v.setAlignment(Pos.CENTER);
    h.setAlignment(Pos.CENTER);
  }

  public AboutDialog(SelfManifest selfManifest) {
    log.atDebug().log("Manifest:{}", selfManifest);
    this.selfManifest = selfManifest;

    val text = buildContent();
    int row = 0;
    for (Map.Entry<String, Supplier<String>> entry : text.entrySet()) {
      val keyLabel = new Label(entry.getKey() + ":");
      val valueLabel = new Label(entry.getValue().get());

      content.addRow(row++, keyLabel, valueLabel);
    }
    getDialogPane().setContent(v);
  }

  @Contract(" -> new")
  private @NotNull Object2ObjectMap<String, Supplier<String>> buildContent() {
    return new Object2ObjectArrayMap<>() {
      {
        put(
            ManifestAttribute.IMPLEMENTATION_TITLE.getAttributeName(),
            selfManifest::implementationTitle);
        put(ManifestAttribute.BUILT_OS.getAttributeName(), selfManifest::builtOS);
        put(ManifestAttribute.BUILT_JDK.getAttributeName(), selfManifest::builtJDK);
        put(ManifestAttribute.BUILT_BY.getAttributeName(), selfManifest::builtBy);
        put(ManifestAttribute.BUILT_HOST.getAttributeName(), selfManifest::builtHost);
        put(
            ManifestAttribute.BUILT_DATE.getAttributeName(),
            () -> selfManifest.builtDate().toString());
        put(
            ManifestAttribute.IMPLEMENTATION_GROUP.getAttributeName(),
            selfManifest::implementationGroup);
        put(
            ManifestAttribute.IMPLEMENTATION_VERSION.getAttributeName(),
            selfManifest::implementationVersion);
        put(ManifestAttribute.AGENT_CLASS.getAttributeName(), selfManifest::agentClass);
        put(ManifestAttribute.MAIN_CLASS.getAttributeName(), selfManifest::mainClass);
        put(ManifestAttribute.PREMAIN_CLASS.getAttributeName(), selfManifest::premainClass);
        put(ManifestAttribute.SCM_BRANCH.getAttributeName(), selfManifest::scmBranch);
        put(ManifestAttribute.SCM_COMMIT_AUTHOR.getAttributeName(), selfManifest::scmCommitAuthor);
        put(ManifestAttribute.SCM_REPOSITORY.getAttributeName(), selfManifest::scmRepository);
        put(ManifestAttribute.SCM_COMMIT_HASH.getAttributeName(), selfManifest::scmCommitHash);
        put(
            ManifestAttribute.SCM_COMMIT_MESSAGE.getAttributeName(),
            selfManifest::scmCommitMessage);
        put(
            ManifestAttribute.SCM_COMMIT_DATE.getAttributeName(),
            () -> selfManifest.scmCommitDate().toString());
      }
    };
  }
}

File: src/main/kotlin/org/fx/debugger/view/AboutDialog.kt
package org.fx.debugger.view

import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.inject.Inject
import jakarta.inject.Singleton
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.scene.control.ButtonType
import javafx.scene.control.Dialog
import javafx.scene.control.Label
import javafx.scene.layout.GridPane
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import lombok.extern.slf4j.Slf4j
import org.fx.debugger.constant.ManifestAttribute
import org.fx.debugger.model.SelfManifest
import org.jetbrains.annotations.Contract

@Singleton
@Slf4j
class AboutDialog @Inject constructor(private val selfManifest: SelfManifest) : Dialog<Void?>() {
  private val log = KotlinLogging.logger {}
  private val content = GridPane()

  init {
    content.padding = Insets(20.0)
    content.hgap = 10.0
    content.vgap = 5.0
  }

  init {
    dialogPane.buttonTypes.add(ButtonType.CLOSE)
    isResizable = true
    title = "About Fx-Debugger"
  }

  private val h = HBox(content).apply { alignment = Pos.CENTER }

  private val v = VBox(h).apply { alignment = Pos.CENTER }

  init {
    val text = buildContent()
    var row = 0
    for ((key, value) in text) {
      val keyLabel = Label("$key:")
      val valueLabel = Label(value)

      content.addRow(row++, keyLabel, valueLabel)
    }
    dialogPane.content = v
  }

  @Contract(" -> new")
  private fun buildContent(): Map<String, String> {
    return mapOf(
        ManifestAttribute.IMPLEMENTATION_TITLE.attributeName to selfManifest.implementationTitle,
        ManifestAttribute.BUILT_OS.attributeName to selfManifest.builtOS)
    //    return object : Object2ObjectArrayMap<String, Supplier<String>>() {
    //      init {
    //        put(
    //          ManifestAttribute.IMPLEMENTATION_TITLE.attributeName,
    //          selfManifest::implementationTitle
    //        )
    //        put(ManifestAttribute.BUILT_OS.attributeName, selfManifest::builtOS)
    //        put(ManifestAttribute.BUILT_JDK.attributeName, selfManifest::builtJDK)
    //        put(ManifestAttribute.BUILT_BY.attributeName, selfManifest::builtBy)
    //        put(ManifestAttribute.BUILT_HOST.attributeName, selfManifest::builtHost)
    //        put(
    //          ManifestAttribute.BUILT_DATE.attributeName,
    //          Supplier { selfManifest.builtDate.toString() })
    //        put(
    //          ManifestAttribute.IMPLEMENTATION_GROUP.attributeName,
    //          selfManifest::implementationGroup
    //        )
    //        put(
    //          ManifestAttribute.IMPLEMENTATION_VERSION.attributeName,
    //          selfManifest::implementationVersion
    //        )
    //        put(ManifestAttribute.AGENT_CLASS.attributeName, selfManifest::agentClass)
    //        put(ManifestAttribute.MAIN_CLASS.attributeName, selfManifest::mainClass)
    //        put(ManifestAttribute.PREMAIN_CLASS.attributeName, selfManifest::premainClass)
    //        put(ManifestAttribute.SCM_BRANCH.attributeName, selfManifest::scmBranch)
    //        put(ManifestAttribute.SCM_COMMIT_AUTHOR.attributeName, selfManifest::scmCommitAuthor)
    //        put(ManifestAttribute.SCM_REPOSITORY.attributeName, selfManifest::scmRepository)
    //        put(ManifestAttribute.SCM_COMMIT_HASH.attributeName, selfManifest::scmCommitHash)
    //        put(
    //          ManifestAttribute.SCM_COMMIT_MESSAGE.attributeName,
    //          selfManifest::scmCommitMessage
    //        )
    //        put(
    //          ManifestAttribute.SCM_COMMIT_DATE.attributeName,
    //          Supplier { selfManifest.scmCommitDate.toString() })
    //      }
    //    }
  }
}
